### PR TITLE
Adds AdSense support

### PR DESCRIPTION
**Pull request by jonkeller**
Adds AdSense support to most relevant audits. First Ad Paint support for AdSense will come in a separate PR, after both this one and #121 go in.

### CODE REVIEW

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js [hunk 1]
```diff
@@ -61,7 +61,7 @@ class LoadsGptFromSgdn extends Audit {
     const networkRecords = await NetworkRecords.request(devtoolsLog, context);
     const gptUrl = networkRecords.map((r) => new URL(r.url)).find(isGptTag);
     if (!gptUrl) {
-      return auditNotApplicable.NoGpt;
+      return auditNotApplicable.NoTag;
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js]: Obviously more work is needed in this file. This is just to get it to continue to compile.

File: lighthouse-plugin-publisher-ads/messages/common-strings.js [hunk 2]
```diff
@@ -25,7 +25,6 @@ const UIStrings = {
   NOT_APPLICABLE__NO_ADS: 'No ads requested',
   NOT_APPLICABLE__NO_BIDS: 'No bids detected',
   NOT_APPLICABLE__NO_EVENT_MATCHING_REQ: 'No event matches network records',
-  NOT_APPLICABLE__NO_GPT: 'GPT not requested',
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/messages/common-strings.js]: Using NOT_APPLICABLE__NO_TAG instead.

File: lighthouse-plugin-publisher-ads/messages/en-US.json [hunk 3]
```diff
@@ -299,20 +299,20 @@
     "message": "GPT tag is loaded from recommended host",
     "description": ""
   },
-  "audits/loads-gpt-over-https.js | description": {
-    "message": "For privacy and security, always load GPT over HTTPS. Insecure pages should explicitly request the GPT script securely. Example:`<script async src=\"https://securepubads.g.doubleclick.net/tag/js/gpt.js\">`. [Learn more](https://developers.google.com/publisher-ads-audits/reference/audits/loads-gpt-over-https).",
+  "audits/loads-ad-tag-over-https.js | description": {
+    "message": "For privacy and security, always load GPT/AdSense over HTTPS. Insecure pages should explicitly request the ad script securely. Example:`<script async src=\"https://securepubads.g.doubleclick.net/tag/js/gpt.js\">` or `<script async src=\"https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js\">`. [Learn more](https://developers.google.com/publisher-ads-audits/reference/audits/loads-ad-tag-over-https).",
```
[the remark this answers is not in the document]
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/messages/en-US.json]: Done.

File: lighthouse-plugin-publisher-ads/messages/locales/en-XL.json [hunk 4]
```diff
@@ -224,16 +224,16 @@
   "audits/loads-gpt-from-sgdn.js | title": {
     "message": "ĜṔT̂ t́âǵ îś l̂óâd́êd́ f̂ŕôḿ r̂éĉóm̂ḿêńd̂éd̂ h́ôśt̂"
   },
-  "audits/loads-gpt-over-https.js | description": {
-    "message": "F̂ór̂ ṕr̂ív̂áĉý âńd̂ śêćûŕît́ŷ, ál̂ẃâýŝ ĺôád̂ ǴP̂T́ ôv́êŕ ĤT́T̂ṔŜ. Ín̂śêćûŕê ṕâǵêś ŝh́ôúl̂d́ êx́p̂ĺîćît́l̂ý r̂éq̂úêśt̂ t́ĥé ĜṔT̂ śĉŕîṕt̂ śêćûŕêĺŷ. Éx̂ám̂ṕl̂é:`<ŝćr̂íp̂t́ âśŷńĉ śr̂ć=\"ĥt́t̂ṕŝ://śêćûŕêṕûb́âd́ŝ.ǵ.d̂óûb́l̂éĉĺîćk̂.ńêt́/t̂áĝ/j́ŝ/ǵp̂t́.ĵś\">`. [L̂éâŕn̂ ḿôŕê](h́t̂t́p̂ś://d̂év̂él̂óp̂ér̂ś.ĝóôǵl̂é.ĉóm̂/ṕûb́l̂íŝh́êŕ-âd́ŝ-áûd́ît́ŝ/ŕêf́êŕêńĉé/âúd̂ít̂ś/l̂óâd́ŝ-ǵp̂t́-ôv́êŕ-ĥt́t̂ṕŝ)."
+  "audits/loads-ad-tag-over-https.js | description": {
```
[the remark this answers is not in the document]
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/messages/locales/en-XL.json]: Done

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 5]
```diff
@@ -94,6 +84,91 @@ function isGptAdRequest(request) {
...
+ * Checks if the url is loading an AdSense script.
+ * @param {URL|string} url
+ * @return {boolean}
+ */
+function isAdSenseTag(url) {
+  const {host, pathname} = toURL(url);
+  const matchesHost = [
+    'pagead2.googlesyndication.com'].includes(host);
+  const matchesPath =
+      ['/pagead/js/adsbygoogle.js'].includes(pathname);
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js]: And `show_ads.js`

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 6]
```diff
@@ -94,6 +84,91 @@ function isGptAdRequest(request) {
...
+  return matchesHost && matchesPath;
+}
+
+/**
+ * Checks if the url is loading an AdSense script. Kind of pointless since
+ * AdSense doesn't have a separate loader, but provided anyway for symmetry with
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js]: AdSense has two loaders! adsbygoogle (modern, preferred) and show_ads.js (deprecated, sync only). The impl is show_ads_impl.js

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 7]
```diff
@@ -197,18 +264,24 @@ function getAbbreviatedUrl(url) {
...
-  isGptAdRequest,
-  hasImpressionPath,
-  isGpt,
   isGptTag,
+  isGptImplTag,
+  isGpt,
+  isGptAdRequest,
+  isGptIframe,
+  isAdSense,
+  isAdSenseTag,
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Maybe put AdSense before GPT, to sort alphabetically

File: lighthouse-plugin-publisher-ads/audits/gpt-bids-parallel.js [hunk 8]
```diff
@@ -73,7 +73,7 @@ class GptBidsInParallel extends Audit {
     const trace = artifacts.traces[Audit.DEFAULT_PASS];
     const network = await NetworkRecords.request(devtoolsLog, context);
 
-    const pubadsImpl = network.find((r) => isImplTag(r.url));
+    const pubadsImpl = network.find((r) => isGptImplTag(r.url));
     if (!pubadsImpl) {
       return auditNotApplicable.NoTag;
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/audits/gpt-bids-parallel.js]: NoGptTag?

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js (hunk 1 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js]: Undoing this change

File: lighthouse-plugin-publisher-ads/messages/common-strings.js (hunk 2 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/messages/common-strings.js]: Undoing this change.

File: lighthouse-plugin-publisher-ads/audits/gpt-bids-parallel.js (hunk 8 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/audits/gpt-bids-parallel.js]: Done

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 7 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Done (but not committed yet)

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 6 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Whoa...I had no idea. Thanks!

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 5 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Done

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 6 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Done

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 7 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: ...and now committed.

File: lighthouse-plugin-publisher-ads/computed/ad-paint-time.js [hunk 9]
```diff
@@ -71,11 +71,17 @@ class LanternAdPaintTime extends AdLanternMetric {
     const adFrameIds = new Set(iframes.map(
       /** @param {Artifacts['IFrameElement']} s */
       (s) => s.frame && s.frame.id));
+    console.log('------- adFrameIds ---------'); // DO NOT SUBMIT - remove before merging PR
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/computed/ad-paint-time.js]: These 6 log/dir statements will be removed before merge...

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 10]
```diff
@@ -37,12 +37,67 @@ function isGoogleAds(url) {
...
+}
+
+/**
+ * @param {Artifacts['IFrameElement']} iframe
+ * @return {boolean}
+ */
+function isAdSenseIframe(iframe) {
+  // TODO(jonkeller) Discuss with warrengm@ which of these is correct
+  // return /(^google_ads_frame)/.test(iframe.id);
+  return /(^aswift_\d+)/.test(iframe.id);
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js]: @warrengm Let's talk sometime today?
Basically the issue is that it's rendered like <iframe id="aswift_\d+">...<iframe id="google_ads_frame\d+" src="somewhere-cross-domain">
In both cases (the line that's commented out here, and the line that isn't) we end up here: https://github.com/googleads/publisher-ads-lighthouse-plugin/blob/e57628a1fb0787bd5eb3d9a004325e50425077b6/lighthouse-plugin-publisher-ads/audits/first-ad-paint.js#L74

File: lighthouse-plugin-publisher-ads/computed/ad-paint-time.js (hunk 9 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/computed/ad-paint-time.js]: Done

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 10 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Talked/fixed.

**Comment by jonkeller** [comment]
Current state:
* Latency of first ad render - this audit is still GPT-only, but should be modified to work with AdSense in a separate PR
* GPT and bids loaded in parallel, and
* Header bidding is parallelized - header bidding is not relevant to AdSense, so these two audits haven't/won't be updated
* GPT tag is loaded from recommended host - this is also not relevant to AdSense, as it has not recently launched any new hostname.

File: lighthouse-plugin-publisher-ads/audits/loads-ad-tag-over-https.js [hunk 11]
```diff
@@ -69,35 +72,35 @@ class LoadsGptOverHttps extends Audit {
...
       return returnVal;
     }
 
     return {
-      numericValue: details.numGptHttpReqs,
-      score: details.numGptHttpReqs ? 0 : 1,
-      displayValue: details.numGptHttpReqs ?
+      numericValue: details.numAdTagHttpReqs,
+      score: details.numAdTagHttpReqs ? 0 : 1,
+      displayValue: details.numAdTagHttpReqs ?
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/audits/loads-ad-tag-over-https.js]: Can you add a TODO to include a details to note which scripts are loaded over http?

I think we can shorten the audit description after we add the details table.

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 12]
```diff
@@ -37,12 +37,65 @@ function isGoogleAds(url) {
...
+ * Checks if the url is loading an AdSense script.
  * @param {URL|string} url
  * @return {boolean}
  */
-function isImplTag(url) {
-  return /(^\/gpt\/pubads_impl_\d+.js)/.test(toURL(url).pathname);
+function isAdSenseTag(url) {
+  const {host, pathname} = toURL(url);
+  const matchesHost = [
+    'pagead2.googlesyndication.com'].includes(host);
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Nit: just check for string equality here and in `isAdSenseImplTag`?

File: lighthouse-plugin-publisher-ads/audits/loads-ad-tag-over-https.js (hunk 11 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/audits/loads-ad-tag-over-https.js]: Done

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 12 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js]: Done

**Comment by jonkeller** [comment]
Fixes #2, with the exception of the First Ad Paint audit.

**Pull request merged**